W-20130488 add context to a2a protocol flow (#545)
mulesoft/docs-code-builder: modules/ROOT/pages/af-agent-networks.adoc

The asset might also be published to Exchange.
. An asset that has been added to this agent network project from Exchange.

== A2A Protocol 

The Agent2Agent (A2A) Protocol governs agent-to-agent communication. This protocol powers orchestration, observability, and governance features in agent networks. MuleSoft supports v0.3.0 of the A2A Protocol Specification.

=== Context and Task ID Scoping in Agent Networks

In MuleSoft agent networks, the brokers receiving a request always generate a `contextId` and `taskId`. These IDs define the state and scope of a specific conversation between two agents. A `taskId` is always matched to a `contextId`, but a `contextId` can exist without a `taskId`.

> In a multi-agent network, a client sends a request to Broker_1 and Broker_1 generates the necessary IDs for that request. When Broker_1 sends a new request to the next broker or non-broker agent in line, that broker or non-broker agent establishes a unique `contextID` and `taskID` for the new request. (Note that it's not mandatory for non-broker agents to generate a `contextId` and `taskId` when receiving requests from a client.)

Reviewer: I would either do away with the "Note" in the parenthesis or make it an actual note. And also remove the parenthesis.

> In a multi-agent network, a client sends a request to Broker_1 and Broker_1 generates the necessary IDs for that request. When Broker_1 sends a new request to the next broker or non-broker agent in line, that broker or non-broker agent establishes a unique `contextID` and `taskID` for the new request. (Note that it's not mandatory for non-broker agents to generate a `contextId` and `taskId` when receiving requests from a client.)

Author: ```suggestion
In a multi-agent network, a client sends a request to Broker_1 and Broker_1 generates the necessary IDs for that request. When Broker_1 sends a new request to the next broker or non-broker agent in line, that broker or non-broker agent establishes a unique `contextID` and `taskID` for the new request. 

NOTE: It is not mandatory for non-broker agents to generate a `contextId` and `taskId` when receiving requests from a client.
```

Consider a network with a client and two brokers (1 and 2).

* The IDs used between the client and Broker_1 are independent of the IDs used between Broker_1 and Broker_2.
* When Broker_1 delegates a task to Broker_2, Broker 2 (acting as a server) generates its own `contextId` and `taskId`.

> *  Broker_1 is responsible for maintaining a mapping between its own upstream `taskId` (used to respond to its client) and the downstream `taskId` it is tracking with Broker_2.

Reviewer: ```suggestion
*  Broker_1 is responsible for maintaining a mapping between its own upstream `taskId` (used to respond to its client) and the downstream `taskId` it's tracking with Broker_2.
```

The asset might also be published to Exchange.
. An asset that has been added to this agent network project from Exchange.

== A2A Protocol 

> For orchestration, observability, and governance, agents use the A2A specification 0.3.0.

Author: Changed this here. It's more direct.

> For orchestration, observability, and governance, agents use the A2A specification 0.3.0.

Reviewer 2: This is not bad, but I'd personally write it as `A2A protocol v0.3.0 is used for all inter-agent communication. All orchestration, observability, and governance feature are built on top of this protocol`

=== Context and Task ID Scoping in Agent Networks

> In MuleSoft agent networks, the server (the agent or broker receiving the request) always generates the `contextId` and `taskId`. These IDs define the state and scope of a specific conversation between two agents.

Reviewer 2: `the server (the agent or broker receiving the request) ` -> brokers

> In MuleSoft agent networks, the server (the agent or broker receiving the request) always generates the `contextId` and `taskId`. These IDs define the state and scope of a specific conversation between two agents.

Author: ```suggestion
In MuleSoft agent networks, the servers (the brokers receiving the request) always generate the `contextId` and `taskId`. These IDs define the state and scope of a specific conversation between two agents.
```

> In MuleSoft agent networks, the servers (the brokers receiving a request) always generate a `contextId` and `taskId`. These IDs define the state and scope of a specific conversation between two agents. A `taskId` is always matched to a `contextId`, but a `contextId` can exist without a `taskId`.

Author: @marianogonzalez updated lines 69-72

> In MuleSoft agent networks, the servers (the brokers receiving a request) always generate a `contextId` and `taskId`. These IDs define the state and scope of a specific conversation between two agents. A `taskId` is always matched to a `contextId`, but a `contextId` can exist without a `taskId`.

Reviewer 2: `the servers (the brokers receiving a request)` -> the brokers. Non brokers servers with different behaviors can exist

> In a multi-agent network, a client sends a request to a broker and the broker generates the necessary IDs. When the broker sends a request to the next broker in line, the broker doesn't send these same IDs. Instead, the broker establishes a unique `contextID` and `taskID` for the request to the next broker. 

Author: @marianogonzalez In the interest of not having a diagram right away, does this work? I tried to simplify it.

> In a multi-agent network, a client sends a request to a broker and the broker generates the necessary IDs. When the broker sends a request to the next broker in line, the broker doesn't send these same IDs. Instead, the broker establishes a unique `contextID` and `taskID` for the request to the next broker. 

Author: ```suggestion
In a multi-agent network, a client sends a request to a broker and the broker generates the necessary IDs. When that broker sends a request to the next broker in line, the broker doesn't send these same IDs. Instead, the broker establishes a unique `contextID` and `taskID` for the request to the next broker. 
```

> In a multi-agent network, a client sends a request to a broker and the broker generates the necessary IDs. When the broker sends a request to the next broker in line, the broker doesn't send these same IDs. Instead, the broker establishes a unique `contextID` and `taskID` for the request to the next broker. 

Reviewer 2: This is still not clear and potentially misleading:

* `Instead, the broker establishes a unique `contextID` and `taskID` for the request to the next broker. ` -> The term `the broker` is being used to refer to both broker, making it unclear which broker generates what
* This can also be the case with non-broker agents. If a broker calls a specialized agent, although it is not forced to generate a context or task id, it very well might, in which case the behavior is the same. Also notice that a taskId is always matched with a contextId, but a contextId can exist without any taskId

> In a multi-agent network, a client sends a request to Broker_1 and Broker_1 generates the necessary IDs for that request. When Broker_1 sends a new request to the next broker or non-broker agent in line, Broker_1 establishes a unique `contextID` and `taskID` for that new request. (Note that it's not mandatory for brokers to generate a `contextId` and `taskId` when making requests to non-broker agents.)

Reviewer 2: `Broker_1 establishes a unique `contextID` and `taskID` ` -> In this case Broker1 is the client. The ids are always generated by the server, which in this case is either the Broker2 or the non broker agent

> In a multi-agent network, a client sends a request to Broker_1 and Broker_1 generates the necessary IDs for that request. When Broker_1 sends a new request to the next broker or non-broker agent in line, Broker_1 establishes a unique `contextID` and `taskID` for that new request. (Note that it's not mandatory for brokers to generate a `contextId` and `taskId` when making requests to non-broker agents.)

Reviewer 2: `(Note that it's not mandatory for brokers to generate a `contextId` and `taskId` when making requests to non-broker agents.)` -> The ids are always generated by the server, never the client. Non broker servers are the ones that are not mandated to generate these ids

Consider a network with a client and two brokers (1 and 2).

* The IDs used between the client and Broker_1 are independent of the IDs used between Broker_1 and Broker_2.

> * When Broker_1 delegates a task to Broker_2, Broker B (acting as a server) generates its own `contextId` and `taskId`.

Reviewer 2: ` When Broker_1 delegates a task to Broker_2, Broker B` -> Should Broker B be replaced with Broker 2?

> * When Broker_1 delegates a task to Broker_2, Broker B (acting as a server) generates its own `contextId` and `taskId`.

Author: ```suggestion
* When Broker_1 delegates a task to Broker_2, Broker 2 (acting as a server) generates its own `contextId` and `taskId`.
```

*  Broker_1 is responsible for maintaining a mapping between its own upstream `taskId` (used to respond to its client) and the downstream `taskId` it is tracking with Broker_2.
* If Broker_2 requires more information (if it returns `status: input-required`), it provides a `contextId` and `taskId` to Broker_1. Broker_1 uses these IDs to provide the requested input to Broker_2. The client never sees Broker_2's internal IDs.

[cols="1,1,1", options="header"]
|===
| Relationship | Role |  Logic
| Client -> Broker_1 | Broker_1 is server | Generates `contextID_1` and `taskID_1` for the client.
| Agent A -> Broker_2 | Broker_2 is server | Generates `contextID_2` and `taskID_2` for Broker_1.
| Network broker | Broker_1 | Broker_1 maps `contextID_1` and `taskID_1` to `contextID_2` and `taskID_2`.
|===

For more information, see https://a2a-protocol.org/v0.3.0/topics/life-of-a-task/#group-related-interactions[Life of a Task - Group Related Interactions].

=== Non-Supported Features

Agent network doesn't support the following features in the A2A protocol.

> * Streaming

Author: @marianogonzalez Per your review, I removed task history and push notification configs, but those are both in the docs as not supported https://docs.mulesoft.com/anypoint-code-builder/af-agent-networks#considerations

Reviewer 2: that documentation is wrong and needs to be fixed too

Author: Ok, I'll fix that.

The asset might also be published to Exchange.
. An asset that has been added to this agent network project from Exchange.

== A2A Protocol 

> Agent networks use the A2A specification 0.3.0. Brokers use the A2A protocol to accept requests to reach other agents.

Reviewer 2: I would rephrase this entire statement. It's not the agent network that uses A2A, as the network also contain other components such as MCP servers and LLMs.

It's also not just the brokers that use A2A, all the agents (broker or not) must be A2A enabled not only to be orchestrated but also to be observed and governed

Author: @marianogonzalez I simplified the sentence.

=== Context and Task ID Scoping in Agent Networks

> In the A2A protocol, the server (the agent receiving the request) always generates the `contextId` and `taskId`. These IDs define the state and scope of a specific conversation between two agents.

Reviewer 2: > In the A2A protocol, the server (the agent receiving the request) always generates the `contextId` and `taskId`.

Not true. Those IDs are always generated by the server, but the server is not required to always generate them. Our brokers DO always generate them, but that's a property of the broker, not the A2A protocol

Author: @marianogonzalez I changed it to read: In MuleSoft agent networks, the server (the agent receiving the request) always generates the `contextId` and `taskId`.

But should we say In MuleSoft agent networks, the broker (the agent receiving the request) always generates the `contextId` and `taskId`.:

> When an agent acts as a broker, IDs do not pass through from the upstream client. Each "hop" in a multi-agent network establishes its own unique session. 

Reviewer 2: > When an agent acts as a broker, IDs do not pass through from the upstream client. 

Several things here:

1. I get what you're trying to say here, but I don't think this will be clear to customers. Maybe an example between 3 agents with a picture would be helpful. 
2. This is not true just for brokers but any three-party A2A communication.
3. We should unifiy the language. The terms conversation, context and session are being used interchengeably and they not necessarilly represent the same thing (e.g: session could be understood as authenticated session). I would stick to the proper A2A term (contextId) or at most conversation

Author: @marianogonzalez In the interest of not having a diagram right away, I tried to simplify the explanation here. Does this work?

Consider a network with a client and two agents (A and B).

> * The IDs used between the client and Agent A are independent of the IDs used between Agent A and Agent B.

Reviewer 2: in this example, we should use the term `broker` instead of `agent`

* When Agent A delegates a task to Agent B, Agent B (acting as a server) generates its own `contextId` and `taskId`.
*  Agent A is responsible for maintaining a mapping between its own upstream `taskId` (used to respond to its client) and the downstream `taskId` it is tracking with Agent B.
* If Agent B requires more information (if it returns `status: input-required`), Agent B provides a `contextId` and `taskId` to Agent A. Agent A uses these IDs to provide the requested input to Agent B. The client never sees Agent B's internal IDs.

[cols="1,1,1", options="header"]
|===
| Relationship | Role |  Logic
| Client -> Agent A | Agent A is server | Generates `contextID_1` and `taskID_1` for the client.
| Agent A -> Agent B | Agent B is server | Generates `contextID_2` and `taskID_2` for Agent A.
| Network broker | Agent A | Agent A maps `contextID_1` and `taskID_1` to `contextID_2` and `taskID_2`.
|===

For more information, see https://a2a-protocol.org/v0.3.0/topics/life-of-a-task/#group-related-interactions[Life of a Task - Group Related Interactions].

=== Non-Supported Features

Agent network doesn't support the following features in the A2A protocol.

> * Task History

Reviewer 2: not true, this is supported

Author: OK, I removed it, but in the docs, it's listed as not available. Assumed that meant not supported. https://docs.mulesoft.com/anypoint-code-builder/af-agent-networks#considerations

> * Push Notifications Config

Reviewer 2: not true, supported

Author: Also listed as unavailable https://docs.mulesoft.com/anypoint-code-builder/af-agent-networks#considerations
